File: src/algorithm/sorting/quick-sort/quickSort.js
#! /usr/local/bin/node


class QuickSort {
    sort(originalArr=[]) {
        let list = [...originalArr];

        if (list.length <= 1) {
            return list;
        }

        let pivot = list.shift();
        let leftList = [];
        let rightList = [];
        let middleList = [pivot];

        for (let i = 0, length = list.length; i < length; i++) {
            let item = list[i];
            if (item <= pivot) {
                leftList.push(item);
            } else {
                rightList.push(item);
            }
        }

        leftList = this.sort(leftList);
        rightList = this.sort(rightList);

        return [...leftList, ...middleList, ...rightList];
    }

    isSorted(array=[]) {
        if (array.length <= 1) {
            return true;
        }

        for (let i = 1; i < array.length; i++) {
            if (array[i] < array[i - 1]) {
                return false;
            }
        }

        return true;
    }

    test() {
        const array = new Array(40000).fill(undefined).map(item => Math.random() * 10000);
        console.time('sort');
        let result = this.sort(array);
        console.timeEnd('sort');
        let sorted = this.isSorted(result);
        console.log('sorted:', sorted);
    }
}

new QuickSort().test();

File: src/algorithm/sorting/quick-sort/quick_sort.py
#! /usr/local/bin/python3


class QuickSort:
    def sort(self, original_list):
        if original_list is None:
            return []

        new_list = original_list.copy()

        if len(new_list) <= 1:
            return new_list

        povit = new_list.pop()
        left_list = []
        right_list = []
        middle_list = [povit]

        for item in new_list:
            if item <= povit:
                left_list.append(item)
            else:
                right_list.append(item)

        left_list = self.sort(left_list)
        right_list = self.sort(right_list)

        return left_list + middle_list + right_list

    def test(self):
        import time

        start = time.process_time()
        print(self.sort([34, 454, 3, 565, 7, 8, 23, 5, 2, 97, 56, 6]))
        end = time.process_time()
        print('sort:', (end - start) * 1000)

if __name__ == '__main__':
    QuickSort().test()
